pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/AvgShedder.java

```diff
@@ -0,0 +1,312 @@
+                double traffic = isMsgRateToOffload
+                        ? shortTermData.getMsgRateIn() + shortTermData.getMsgRateOut()
+                        : shortTermData.getMsgThroughputIn() + shortTermData.getMsgThroughputOut();
+                return Pair.of(e, traffic);
+            }).sorted((e1, e2) ->
+                    Double.compare(e2.getRight(), e1.getRight())
+            ).forEach(e -> {
+                Map.Entry<String, BundleData> bundle = e.getLeft();
+                double traffic = e.getRight();
+                if (traffic > 0 && traffic <= trafficMarkedToOffload.getValue()) {
+                    selectedBundlesCache.put(overloadedBroker, bundle.getKey());
+                    bundleBrokerMap.put(bundle.getValue(), underloadedBroker);
+                    trafficMarkedToOffload.add(-traffic);
+                    if (log.isDebugEnabled()) {
+                        log.debug("Found bundle to unload:{}, isMsgRateToOffload:{}, traffic:{}",
+                                bundle, isMsgRateToOffload, traffic);
+                    }
+                }
+            });
+        }
+        return selectedBundlesCache;
+    }
+
+    @Override
+    public void onActiveBrokersChange(Set<String> activeBrokers) {
+        LoadSheddingStrategy.super.onActiveBrokersChange(activeBrokers);
+    }
+
+    private Double calculateScores(String broker, LocalBrokerData localBrokerData, final ServiceConfiguration conf) {
+        double score = localBrokerData.getMaxResourceUsageWithWeight(
+                conf.getLoadBalancerCPUResourceWeight(),
+                conf.getLoadBalancerDirectMemoryResourceWeight(),
+                conf.getLoadBalancerBandwithInResourceWeight(),
+                conf.getLoadBalancerBandwithOutResourceWeight()) * 100;
+        return score;
+    }
+
+    @Override
+    public Optional<String> selectBroker(Set<String> candidates, BundleData bundleToAssign, LoadData loadData,
+                                         ServiceConfiguration conf) {
+        if (!bundleBrokerMap.containsKey(bundleToAssign) || !candidates.contains(bundleBrokerMap.get(bundleToAssign))) {
+            // cluster initializing or broker is shutdown
+            if (log.isDebugEnabled()) {
+                if (!bundleBrokerMap.containsKey(bundleToAssign)) {
+                    log.debug("cluster is initializing");
+                } else {
+                    log.debug("expected broker:{} is shutdown, candidates:{}", bundleBrokerMap.get(bundleToAssign),
+                            candidates);
+                }
+            }
+            String broker = getExpectedBroker(candidates, bundleToAssign);
+            bundleBrokerMap.put(bundleToAssign, broker);
+            return Optional.of(broker);
+        } else {
+            return Optional.of(bundleBrokerMap.get(bundleToAssign));
+        }
+    }
+
+    private static long hash(String key) {
```

Comment:
We have only one place to use this method, I think we may not need to create a hash method for it.